{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b7cafa33-bed2-4407-87f5-51d133cec1df",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/4TB/TCGA_Colorectal/scripts/3-processing'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "055d1b4c-1d1d-40f9-8c0e-070ee4143634",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from guppy import hpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3dbadbb2-d44a-4e31-87e0-251dd237d88c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "start\n",
      "Partition of a set of 604766 objects. Total size = 123713725 bytes.\n",
      " Index  Count   %     Size   % Cumulative  % Kind (class / dict of class)\n",
      "     0      1   0 43824841  35  43824841  35 pandas.core.frame.DataFrame\n",
      "     1 162823  27 21354406  17  65179247  53 str\n",
      "     2 148504  25 12485272  10  77664519  63 tuple\n",
      "     3     83   0  6059214   5  83723733  68 numpy.ndarray\n",
      "     4  14919   2  5262848   4  88986581  72 dict (no owner)\n",
      "     5  66008  11  5015670   4  94002251  76 bytes\n",
      "     6  33471   6  4843680   4  98845931  80 types.CodeType\n",
      "     7  30453   5  4385232   4 103231163  83 function\n",
      "     8   4037   1  3813448   3 107044611  87 type\n",
      "     9   1450   0  2279144   2 109323755  88 dict of module\n",
      "<1436 more rows. Type e.g. '_.more' to view.>\n",
      "end\n"
     ]
    }
   ],
   "source": [
    "print(\"start\")\n",
    "import sys, os\n",
    "from os.path import join as pj\n",
    "import pandas as pd\n",
    "\n",
    "# ====== PROJECT SCAFFOLD ======\n",
    "# Arguments passed into script by Docker command\n",
    "project = sys.argv[1]\n",
    "file_id = sys.argv[2]\n",
    "\n",
    "# Mounted volume\n",
    "wd = '/mnt/4TB/TCGA_Colorectal/scripts/3-processing'\n",
    "beep = pj(wd, '43cbb4c9-4dc8-4e7a-ac69-779f5a76dced.pq')\n",
    "\n",
    "df = pd.read_parquet(beep)\n",
    "\n",
    "# Variant must be mapped to a gene\n",
    "df = df[df['VEP_SYMBOL']!='']\n",
    "\n",
    "# Fill empty AF with 0\n",
    "df['VEP_gnomADg_AF'].replace('',0,inplace=True)\n",
    "\n",
    "# Fill empty with 1\n",
    "df['VEP_am_pathogenicity'].replace('',1,inplace=True)\n",
    "\n",
    "df = df.astype({\n",
    "    'VEP_gnomADg_AF':'float64',\n",
    "    'VEP_am_pathogenicity':'float64',\n",
    "    'FORMAT_GQ':'int64',\n",
    "    'FORMAT_DP':'int64',\n",
    "})\n",
    "\n",
    "# Need to adjust pipeline to handle allosomal dosage and homologous PAR\n",
    "df = df[(df['CHROM']!='chrX') & (df['CHROM']!='chrY')]\n",
    "\n",
    "\n",
    "df = df[(df['VEP_IMPACT']=='HIGH') | (df['VEP_IMPACT']=='MODERATE')]\n",
    "\n",
    "# 99% of of low quality (GQ<10) variants are very low depth (DP<5)\n",
    "# ACMG recommends depth 10, but there are a decent amount high GQ variants between DP 5-10\n",
    "# It would be hard to justify a high-priority feature with very low depth\n",
    "df = df[df['FORMAT_DP']>=5]\n",
    "\n",
    "\n",
    "# All of the non-missense will be weighted by allele freq as planned\n",
    "# The MODERATE-impact indels are as rare as HIGH-impact frameshift mutations, and \"small\" indels can be 1,000 bp long\n",
    "# There is a lot of evidence suggesting 0.1% is the cutoff, and much of the high-impact variation is common\n",
    "# ^ https://docs.google.com/spreadsheets/d/1E7ZQrYo5kXu57rRl9zaSyu3W11cqb_qLjRuc309JMXg/edit#gid=1966751467\n",
    "\n",
    "# AlphaMissense claims to accounts for not only structure, but also human/primate evolutionary conservation (rare/common AF)\n",
    "# However, there are loads of extremely high AF (0.20-1.0) variants marked as pathogenic\n",
    "afs     = df['VEP_gnomADg_AF'].tolist()\n",
    "qc_afs  = []\n",
    "for af in afs:\n",
    "    qc_af = (1-af)**40\n",
    "    qc_af = round(qc_af, 5)\n",
    "    qc_afs.append(qc_af)\n",
    "df['QC_AF'] = qc_afs\n",
    "\n",
    "\n",
    "\n",
    "# GQ is a phred score = https://gatk.broadinstitute.org/hc/en-us/articles/360035531872\n",
    "# We want it's likelihood 0.0-1.0\n",
    "gqs    = df['FORMAT_GQ'].tolist()\n",
    "qc_gqs = []\n",
    "for gq in gqs:\n",
    "    qc_gq = 1-10**-(gq/10)\n",
    "    qc_gqs.append(qc_gq)\n",
    "df['QC_GQ'] = qc_gqs\n",
    "\n",
    "\n",
    "# PL docs = https://gatk.broadinstitute.org/hc/en-us/articles/360035890451\n",
    "gts    = df['FORMAT_GT'].tolist()\n",
    "pls    = df['FORMAT_PL'].tolist()\n",
    "gqs    = df['QC_GQ'].tolist()\n",
    "qc_gts = []\n",
    "for e, gt in enumerate(gts):\n",
    "    # Either way, it's a hom\n",
    "    if (gt=='1/2'): qc_gt=2.0\n",
    "    \n",
    "    # There is a chance that this hom is actually a het\n",
    "    elif (gt=='1/1'): qc_gt=2.0-(1-gqs[e])\n",
    "    \n",
    "    # Does it lean toward hom or no variation? Check its PL values\n",
    "    # Note that `1/2` has 6 PL values, so making PL a column would be tricky\n",
    "    elif (gt=='0/1'):\n",
    "        pl = pls[e].split(',')\n",
    "        pl0, pl2 = pl[0], pl[2]\n",
    "        \n",
    "        # e.g. [2,0,2] doesn't lean either way, so assume it's a het\n",
    "        if (pl0==pl2): qc_gt=1.0\n",
    "            \n",
    "        # Otherwise, does uncertainty lean toward either hom or no variation?\n",
    "        else:\n",
    "            pl0, pl2 = int(pl0), int(pl2)\n",
    "            # e.g. [8276,0,6] chance the variant is a hom\n",
    "            if (pl0>pl2): qc_gt=2-gqs[e]\n",
    "            #e.g. [5,0,165] chance the variant does not exist\n",
    "            else: qc_gt=gqs[e]\n",
    "    else:\n",
    "        msg = f\"Error - unknown genotype:{gt}\"\n",
    "    qc_gts.append(qc_gt)\n",
    "df['QC_GT'] = qc_gts\n",
    "\n",
    "# Compile it into a weight column\n",
    "df['QC_Weight'] = df['QC_AF']*df['QC_GT']\n",
    "\n",
    "df.reset_index(inplace=True,drop=True)\n",
    "\n",
    "h = hpy()\n",
    "print(h.heap())\n",
    "\n",
    "# df.to_parquet(efs_qc)\n",
    "print(\"end\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6f6f9726-5b53-4932-b451-53ff1fc8ee31",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>VEP_Allele</th>\n",
       "      <th>VEP_Consequence</th>\n",
       "      <th>VEP_IMPACT</th>\n",
       "      <th>...</th>\n",
       "      <th>INFO_MQ</th>\n",
       "      <th>INFO_QD</th>\n",
       "      <th>INFO_SOR</th>\n",
       "      <th>INFO_BaseQRankSum</th>\n",
       "      <th>INFO_MQRankSum</th>\n",
       "      <th>INFO_ReadPosRankSum</th>\n",
       "      <th>QC_AF</th>\n",
       "      <th>QC_GQ</th>\n",
       "      <th>QC_GT</th>\n",
       "      <th>QC_Weight</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69428</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>10861.06</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>...</td>\n",
       "      <td>29.56</td>\n",
       "      <td>26.75</td>\n",
       "      <td>1.150</td>\n",
       "      <td>-2.239</td>\n",
       "      <td>1.312</td>\n",
       "      <td>-0.598</td>\n",
       "      <td>0.79628</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.59256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69511</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1479.06</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>...</td>\n",
       "      <td>27.90</td>\n",
       "      <td>27.91</td>\n",
       "      <td>5.709</td>\n",
       "      <td>2.569</td>\n",
       "      <td>-5.146</td>\n",
       "      <td>0.818</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>953279</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>3604.06</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>...</td>\n",
       "      <td>56.39</td>\n",
       "      <td>30.03</td>\n",
       "      <td>1.169</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>976215</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>335.64</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>...</td>\n",
       "      <td>60.00</td>\n",
       "      <td>3.90</td>\n",
       "      <td>1.312</td>\n",
       "      <td>-1.583</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.024</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>978953</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>346.64</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>...</td>\n",
       "      <td>60.00</td>\n",
       "      <td>12.84</td>\n",
       "      <td>1.179</td>\n",
       "      <td>2.656</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.753</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 75 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM     POS ID REF ALT      QUAL FILTER VEP_Allele   VEP_Consequence  \\\n",
       "0  chr1   69428  .   T   G  10861.06      .          G  missense_variant   \n",
       "1  chr1   69511  .   A   G   1479.06      .          G  missense_variant   \n",
       "2  chr1  953279  .   T   C   3604.06      .          C  missense_variant   \n",
       "3  chr1  976215  .   A   G    335.64      .          G  missense_variant   \n",
       "4  chr1  978953  .   C   G    346.64      .          G  missense_variant   \n",
       "\n",
       "  VEP_IMPACT  ... INFO_MQ INFO_QD INFO_SOR INFO_BaseQRankSum INFO_MQRankSum  \\\n",
       "0   MODERATE  ...   29.56   26.75    1.150            -2.239          1.312   \n",
       "1   MODERATE  ...   27.90   27.91    5.709             2.569         -5.146   \n",
       "2   MODERATE  ...   56.39   30.03    1.169              None           None   \n",
       "3   MODERATE  ...   60.00    3.90    1.312            -1.583          0.000   \n",
       "4   MODERATE  ...   60.00   12.84    1.179             2.656          0.000   \n",
       "\n",
       "  INFO_ReadPosRankSum    QC_AF QC_GQ QC_GT QC_Weight  \n",
       "0              -0.598  0.79628   1.0   2.0   1.59256  \n",
       "1               0.818  0.00000   1.0   2.0   0.00000  \n",
       "2                None  0.00000   1.0   2.0   0.00000  \n",
       "3              -0.024  0.00000   1.0   1.0   0.00000  \n",
       "4               0.753  0.00000   1.0   1.0   0.00000  \n",
       "\n",
       "[5 rows x 75 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
